File: tests/sage_linear.cpp
// Copyright 2020 Marcel Wagenländer

#include "sage_linear.hpp"
#include "chunking.hpp"
#include "cuda_helper.hpp"
#include "helper.hpp"
#include "tensors.hpp"

#include "catch2/catch.hpp"
#include <iostream>

const std::string home = std::getenv("HOME");
const std::string dir_path = home + "/gpu_memory_reduction/alzheimer/data";
const std::string flickr_dir_path = dir_path + "/flickr";
const std::string test_dir_path = dir_path + "/tests";


int test_sage_linear(Matrix<float> *input_self, Matrix<float> *input_neigh, Matrix<float> *in_gradients) {
    std::string path;

    CudaHelper cuda_helper;
    SageLinear sage_linear_layer(&cuda_helper, input_self->num_columns_, in_gradients->num_columns_, input_self->num_rows_);

    Matrix<float> *result = sage_linear_layer.forward(input_self, input_neigh);

    SageLinearGradients *gradients = sage_linear_layer.backward(in_gradients);

    path = test_dir_path + "/result.npy";
    save_npy_matrix(result, path);
    save_params(sage_linear_layer.get_parameters());
    save_grads(gradients, sage_linear_layer.get_gradients());

    char command[] = "/home/ubuntu/gpu_memory_reduction/pytorch-venv/bin/python3 /home/ubuntu/gpu_memory_reduction/alzheimer/tests/sage_linear.py";
    system(command);

    path = test_dir_path + "/value.npy";
    return read_return_value(path);
}

int test_sage_linear_chunked(SageLinearChunkedParent *linear, long chunk_size) {
    std::string path;
    int num_nodes = 1 << 15;
    int num_in_features = 1 << 9;
    int num_out_features = 1 << 8;

    Matrix<float> input_self(num_nodes, num_in_features, false);
    input_self.set_random_values();
    path = test_dir_path + "/input_self.npy";
    save_npy_matrix(&input_self, path);

    Matrix<float> input_neigh(num_nodes, num_in_features, false);
    input_neigh.set_random_values();
    path = test_dir_path + "/input_neigh.npy";
    save_npy_matrix(&input_neigh, path);

    Matrix<float> incoming_gradients(num_nodes, num_out_features, false);
    incoming_gradients.set_random_values();
    path = test_dir_path + "/in_gradients.npy";
    save_npy_matrix(&incoming_gradients, path);

    long num_chunks = ceil((float) num_nodes / (float) chunk_size);
    std::vector<Matrix<float>> input_self_chunked(num_chunks);
    std::vector<Matrix<float>> input_neigh_chunked(num_chunks);
    std::vector<Matrix<float>> incoming_gradients_chunked(num_chunks);
    chunk_up(&input_self, &input_self_chunked, chunk_size);
    chunk_up(&input_neigh, &input_neigh_chunked, chunk_size);
    chunk_up(&incoming_gradients, &incoming_gradients_chunked, chunk_size);

    CudaHelper cuda_helper;
    linear->set(&cuda_helper, num_in_features, num_out_features, chunk_size, num_nodes);

    std::vector<Matrix<float>> *activations = linear->forward(&input_self_chunked, &input_neigh_chunked);

    SageLinearGradientsChunked *gradients = linear->backward(&incoming_gradients_chunked);

    Matrix<float> activations_one(num_nodes, num_out_features, false);
    stitch(activations, &activations_one);

    path = test_dir_path + "/result.npy";
    save_npy_matrix(&activations_one, path);
    save_params(linear->get_parameters());
    save_grads(gradients, linear->get_gradients(), num_nodes);

    char command[] = "/home/ubuntu/gpu_memory_reduction/pytorch-venv/bin/python3 /home/ubuntu/gpu_memory_reduction/alzheimer/tests/sage_linear.py";
    system(command);

    path = test_dir_path + "/value.npy";
    return read_return_value(path);
}

TEST_CASE("SageLinear", "[sagelinear]") {
    std::string path;
    int rows = 1 << 15;
    int columns = 1 << 9;
    int num_out_features = 1 << 8;

    Matrix<float> input_self(rows, columns, true);
    input_self.set_random_values();
    path = test_dir_path + "/input_self.npy";
    save_npy_matrix(&input_self, path);

    Matrix<float> input_neigh(rows, columns, true);
    input_neigh.set_random_values();
    path = test_dir_path + "/input_neigh.npy";
    save_npy_matrix(&input_neigh, path);

    Matrix<float> in_gradients(rows, num_out_features, true);
    in_gradients.set_random_values();
    path = test_dir_path + "/in_gradients.npy";
    save_npy_matrix(&in_gradients, path);

    CHECK(test_sage_linear(&input_self, &input_neigh, &in_gradients));
}

TEST_CASE("SageLinear, chunked", "[sagelinear][chunked]") {
    SageLinearChunked linear;
    CHECK(test_sage_linear_chunked(&linear, 1 << 16));
    CHECK(test_sage_linear_chunked(&linear, 1 << 12));
    CHECK(test_sage_linear_chunked(&linear, 1 << 8));
}

TEST_CASE("SageLinear, pipelined", "[sagelinear][pipelined]") {
    SageLinearPipelined linear;
    CHECK(test_sage_linear_chunked(&linear, 1 << 16));
    CHECK(test_sage_linear_chunked(&linear, 1 << 12));
    CHECK(test_sage_linear_chunked(&linear, 1 << 8));
}

File: tests/sage_linear.py
import numpy as np
import os
import scipy.io
import scipy.sparse as sp
import torch
from helper import (check_equal, check_isclose, save_return_value,
                    write_equal, check_nans, print_small, print_close_equal)


def test_sage_linear():
    all_close = 1.0
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    home = os.getenv("HOME")
    dir_path = home + "/gpu_memory_reduction/alzheimer/data"
    flickr_dir_path = dir_path + "/flickr"
    test_dir_path = dir_path + "/tests"

    path = test_dir_path + "/input_self.npy"
    input_self = np.load(path)
    path = test_dir_path + "/input_neigh.npy"
    input_neigh = np.load(path)
    path = test_dir_path + "/in_gradients.npy"
    in_gradients = np.load(path)

    check_nans(input_self, "Input for self")
    check_nans(input_neigh, "Input for neighbourhood")
    check_nans(in_gradients, "Incoming gradients")

    input_self_torch = torch.from_numpy(input_self)
    input_self_torch = input_self_torch.to(device)
    input_self_torch.requires_grad_()
    input_self_torch.retain_grad()
    input_neigh_torch = torch.from_numpy(input_neigh)
    input_neigh_torch = input_neigh_torch.to(device)
    input_neigh_torch.requires_grad_()
    input_neigh_torch.retain_grad()
    in_gradients_torch = torch.from_numpy(in_gradients)
    in_gradients_torch = in_gradients_torch.to(device)

    path = test_dir_path + "/self_weight.npy"
    self_weight = np.load(path)
    path = test_dir_path + "/self_bias.npy"
    self_bias = np.load(path)
    path = test_dir_path + "/neigh_weight.npy"
    neigh_weight = np.load(path)
    path = test_dir_path + "/neigh_bias.npy"
    neigh_bias = np.load(path)
    path = test_dir_path + "/result.npy"
    sage_linear_result = np.load(path)

    check_nans(self_weight, "Self weight")
    check_nans(self_bias, "Self bias")
    check_nans(neigh_weight, "Neighbourhood weight")
    check_nans(neigh_bias, "Neighbourhood bias")
    check_nans(sage_linear_result, "SageLinear forward")

    self_weight_torch = torch.from_numpy(self_weight)
    self_weight_torch = self_weight_torch.to(device)
    self_weight_torch.requires_grad_()
    self_weight_torch.retain_grad()
    self_bias_torch = torch.from_numpy(self_bias)
    self_bias_torch = self_bias_torch.to(device)
    self_bias_torch.requires_grad_()
    self_bias_torch.retain_grad()

    neigh_weight_torch = torch.from_numpy(neigh_weight)
    neigh_weight_torch = neigh_weight_torch.to(device)
    neigh_weight_torch.requires_grad_()
    neigh_weight_torch.retain_grad()
    neigh_bias_torch = torch.from_numpy(neigh_bias)
    neigh_bias_torch = neigh_bias_torch.to(device)
    neigh_bias_torch.requires_grad_()
    neigh_bias_torch.retain_grad()

    self_result_torch = torch.matmul(input_self_torch, self_weight_torch) + self_bias_torch.T
    neigh_result_torch = torch.matmul(input_neigh_torch, neigh_weight_torch) + neigh_bias_torch.T
    true_sage_linear_result_torch = self_result_torch + neigh_result_torch

    true_sage_linear_result = true_sage_linear_result_torch.detach().cpu().numpy()

    ratio_close = check_isclose(sage_linear_result, true_sage_linear_result)
    ratio_equal = check_equal(sage_linear_result, true_sage_linear_result)
    print_close_equal("SageLinear forward", ratio_close, ratio_equal)
    all_close = all_close * ratio_close

    # BACKPROPAGATION
    true_sage_linear_result_torch.backward(in_gradients_torch)

    path = test_dir_path + "/self_grads.npy"
    self_grads = np.load(path)
    path = test_dir_path + "/neigh_grads.npy"
    neigh_grads = np.load(path)

    path = test_dir_path + "/self_weight_grads.npy"
    self_weight_grads = np.load(path)
    path = test_dir_path + "/self_bias_grads.npy"
    self_bias_grads = np.load(path)
    path = test_dir_path + "/neigh_weight_grads.npy"
    neigh_weight_grads = np.load(path)
    path = test_dir_path + "/neigh_bias_grads.npy"
    neigh_bias_grads = np.load(path)

    true_self_grads = input_self_torch.grad.cpu().numpy()
    true_neigh_grads = input_neigh_torch.grad.cpu().numpy()
    true_self_weight_grads = self_weight_torch.grad.cpu().numpy()
    true_self_bias_grads = self_bias_torch.grad.cpu().numpy()
    true_neigh_weight_grads = neigh_weight_torch.grad.cpu().numpy()
    true_neigh_bias_grads = neigh_bias_torch.grad.cpu().numpy()

    ratio_close = check_isclose(self_grads, true_self_grads)
    ratio_equal = check_equal(self_grads, true_self_grads)
    print("Linear self input: Close: {}, Equal: {}".format(ratio_close, ratio_equal))
    all_close = all_close * ratio_close

    ratio_close = check_isclose(neigh_grads, true_neigh_grads)
    ratio_equal = check_equal(neigh_grads, true_neigh_grads)
    print("Linear neigh input: Close: {}, Equal: {}".format(ratio_close, ratio_equal))
    all_close = all_close * ratio_close

    ratio_close = check_isclose(self_weight_grads, true_self_weight_grads)
    ratio_equal = check_equal(self_weight_grads, true_self_weight_grads)
    print("Linear self weight: Close: {}, Equal: {}".format(ratio_close, ratio_equal))
    all_close = all_close * ratio_close

    ratio_close = check_isclose(self_weight_grads, true_self_weight_grads, True)
    print("Linear self weight: Loose: {}".format(ratio_close))

    ratio_close = check_isclose(self_bias_grads, true_self_bias_grads)
    ratio_equal = check_equal(self_bias_grads, true_self_bias_grads)
    print("Linear self bias: Close: {}, Equal: {}".format(ratio_close, ratio_equal))
    all_close = all_close * ratio_close

    ratio_close = check_isclose(self_bias_grads, true_self_bias_grads, True)
    print("Linear self bias: Loose {}".format(ratio_close))

    ratio_close = check_isclose(neigh_weight_grads, true_neigh_weight_grads)
    ratio_equal = check_equal(neigh_weight_grads, true_neigh_weight_grads)
    print("Linear neigh weight: Close: {}, Equal: {}".format(ratio_close, ratio_equal))
    all_close = all_close * ratio_close

    ratio_close = check_isclose(neigh_weight_grads, true_neigh_weight_grads, True)
    print("Linear neigh weight: Loose: {}".format(ratio_close))

    ratio_close = check_isclose(neigh_bias_grads, true_neigh_bias_grads)
    ratio_equal = check_equal(neigh_bias_grads, true_neigh_bias_grads)
    print("Linear neigh bias: Close: {}, Equal: {}".format(ratio_close, ratio_equal))
    all_close = all_close * ratio_close

    ratio_close = check_isclose(neigh_bias_grads, true_neigh_bias_grads, True)
    print("Linear neigh bias: Loose: {}".format(ratio_close))

    path = test_dir_path + "/value.npy"
    write_equal(all_close, path)


if __name__ == "__main__":
    test_sage_linear()
